cmd/cstorvolumeclaim/controller.go

```diff
@@ -385,3 +399,212 @@ func BaseLabels(cvc *apis.CStorVolumeClaim) map[string]string {
 	}
 	return base
 }
+
+// cvcNeedResize returns true is a cvc requests a resize operation.
+func (c *CVCController) cvcNeedResize(cvc *apis.CStorVolumeClaim) bool {
+	// only bound cstorvolumeclaim can be resized
+	if cvc.Status.Phase != apis.CStorVolumeClaimPhaseBound {
+		return false
+	}
+
+	cv, err := c.clientset.OpenebsV1alpha1().CStorVolumes(cvc.Namespace).
+		Get(cvc.Name, metav1.GetOptions{})
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("falied to get cv %s: %v", cvc.Name, err))
+		return false
+	}
+
+	requestCVCSize := cvc.Spec.Capacity[v1.ResourceStorage]
+	actualCVCSize := cvc.Status.Capacity[v1.ResourceStorage]
+	if requestCVCSize.Cmp(actualCVCSize) > 0 {
+		return true
+	}
+
+	actualCVSize := cv.Spec.Capacity
+	return requestCVCSize.Cmp(actualCVSize) > 0
+}
+
+// resizeCVC will:
+// 1. Mark cvc as resizing.
+// 2. Resize the cstorvolume object.
+// 3. Mark cvc as resizing finished
+func (c *CVCController) resizeCVC(cvc *apis.CStorVolumeClaim) error {
+	if updatedCVC, err := c.markCVCResizeInProgress(cvc); err != nil {
+		glog.Errorf("mark cvc %q as resizing failed: %v", cvc.Name, err)
+		return err
+	} else if updatedCVC != nil {
+		cvc = updatedCVC
+	}
+
+	// Record an event to indicate that cvc-controller is resizing this volume.
+	c.recorder.Event(cvc, corev1.EventTypeNormal, string(apis.CStorVolumeClaimResizing),
+		fmt.Sprintf("CVCController is resizing volume %s", cvc.Name))
+
+	cv, err := c.clientset.OpenebsV1alpha1().CStorVolumes(cvc.Namespace).
+		Get(cvc.Name, metav1.GetOptions{})
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("falied to get cv %s: %v", cvc.Name, err))
+		return err
+	}
+
+	err = func() error {
+		err = c.resizeVolume(cvc, cv)
+		if err != nil {
+			return err
+		}
+		// Resize volume succeeded mark it as resizing finished.
+		return c.markCVCResizeFinished(cvc)
```

Comment:
Instead of marking resize successful we will wait for nth reconciliation and decide based resize status based on cv status capacity.